File: Solutions/797. All Paths From Source to Target/797. All Paths From Source to Target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& g) {
        vector<vector<int>> res;
        vector<int> path;
        path.emplace_back(0);
        dfs(g,res,path,0);
        return res;
    }
    void dfs(vector<vector<int>>& g,vector<vector<int>>& res,vector<int>& path,int node){
        if(node==g.size()-1)
            res.push_back(path);
        for(auto i:g[node]){
            path.emplace_back(i);
            dfs(g,res,path,i);
            path.pop_back();
        }
    }
};

File: Solutions/797. All Paths From Source to Target/797. All Paths From Source to Target.go
func allPathsSourceTarget(g [][]int) [][]int {
    var res [][]int
    var path []int
    path=append(path,0)
    dfs(g,&res,path,0)
    return res
}

func dfs(g [][]int,res *[][]int,path []int,node int){
    if node==len(g)-1{
        tmp:=make([]int,len(path))
        copy(tmp,path)
        *res=append(*res,tmp)
    }
    for _,i:=range g[node]{
        path=append(path,i)
        dfs(g,res,path,i)
        path=path[:len(path)-1]   
    }
}